core/src/test/scala/unit/kafka/server/KafkaApisTest.scala

```diff
@@ -676,37 +691,274 @@ class KafkaApisTest {
       anyInt,
       anyInt,
       anyString,
-      EasyMock.eq(protocols.map(protocol => (protocol.name, protocol.metadata))),
+      EasyMock.capture(capturedProtocols),
       anyObject()
     ))
 
+    EasyMock.replay(groupCoordinator)
+
     createKafkaApis().handleJoinGroupRequest(
       buildRequest(
         new JoinGroupRequest.Builder(
           new JoinGroupRequestData()
             .setGroupId("test")
             .setMemberId("test")
             .setProtocolType("consumer")
-            .setProtocols(new JoinGroupRequestData.JoinGroupRequestProtocolCollection(protocols.iterator.asJava))
-        )
-      )._2)
+            .setProtocols(new JoinGroupRequestData.JoinGroupRequestProtocolCollection(
+              protocols.map { case (name, protocol) => new JoinGroupRequestProtocol()
+                .setName(name).setMetadata(protocol)
+              }.iterator.asJava))
+        ).build()
+      ))
 
-    EasyMock.replay(groupCoordinator)
+    EasyMock.verify(groupCoordinator)
+
+    val capturedProtocolsList = capturedProtocols.getValue
+    assertEquals(protocols.size, capturedProtocolsList.size)
+    protocols.zip(capturedProtocolsList).foreach { case ((expectedName, expectedBytes), (name, bytes)) =>
+      assertEquals(expectedName, name)
+      assertArrayEquals(expectedBytes, bytes)
+    }
+  }
+
+  @Test
+  def testJoinGroupProtocolType(): Unit = {
+    for (version <- ApiKeys.JOIN_GROUP.oldestVersion to ApiKeys.JOIN_GROUP.latestVersion) {
+      testJoinGroupProtocolType(version.asInstanceOf[Short])
+    }
+  }
+
+  def testJoinGroupProtocolType(version: Short): Unit = {
+    EasyMock.reset(groupCoordinator, clientRequestQuotaManager, requestChannel)
+
+    val capturedResponse = expectNoThrottling()
+
+    val groupId = "group"
+    val memberId = "member1"
+    val protocolType = "consumer"
+    val protocols = List(
+      new JoinGroupRequestProtocol().setName("first").setMetadata("first".getBytes()),
+      new JoinGroupRequestProtocol().setName("second").setMetadata("second".getBytes())
+    )
+
+    val capturedCallback = EasyMock.newCapture[JoinGroupCallback]()
+
+    EasyMock.expect(groupCoordinator.handleJoinGroup(
+      anyString,
```

Comment:
Would it be worth adding explicit matchers for these?

Reply:
Good point. Being explicit is way better for sure.